### PR TITLE
[Feat] add omni sleepmode and ack protocol

### PR DESCRIPTION
**Pull request by Flink-ddd**
## Purpose
This PR implements Omni Sleep Mode (Tiered Memory Orchestration) for both NVIDIA and AMD platforms, as proposed in [RFC #1316](https://github.com/vllm-project/vllm-omni/issues/1316).

Key enhancements include:

Tiered Offloading Logic: Support for Level 1 (Weight offloading) and Level 2 (Full de-mapping) sleep stages.

Hardware Abstraction Layer: Unified VRAM auditing and reclamation logic across CUDA and ROCm.

Deterministic Orchestration: Ensuring physical memory release before co-located task execution to prevent OOM.

## Test Plan

1.
Six unit test classes were run on both AMD and NVIDIA systems. The test class is: tests/entrypoints/test_omni_sleep_mode.py. The test scenarios combine ACK signals, LLM and generation, and the sleep and wake-up states of Diffusion.

Especially the following points:
Unit Test 4: Inference consistency and bit-level precision verification after Diffusion wake-up
Unit Test 6: full-cycle audit of Diffusion memory lifecycle.

2. 
The unit test 6 was conducted on different platforms to verify the accuracy and usability of the Diffusion model throughout its entire lifecycle, from Active to sleep to wakeup.

## Test Result

1. 
NVIDIA A6000 TP = 2 （Core pytest test output）

```
tests/entrypoints/test_omni_sleep_mode.py::TestOmniSleepMode::test_diffusion_vram_lifecycle_audit 
=== PRE-TEST GPU CLEANUP ===
GPU cleanup disabled
INFO 02-24 09:58:39 [scheduler.py:229] Chunked prefill is enabled with max_num_batched_tokens=2048.
--- Running test: test_diffusion_vram_lifecycle_audit
INFO 02-24 09:58:39 [omni.py:117] Initializing stages for model: black-forest-labs/FLUX.2-klein-4B
..................................................
[Stage-0] INFO 02-24 09:59:08 [async_omni_diffusion.py:113] AsyncOmniDiffusion initialized with model: black-forest-labs/FLUX.2-klein-4B
INFO 02-24 09:59:08 [omni.py:334] [AsyncOrchestrator] Stage-0 reported ready
INFO 02-24 09:59:08 [omni.py:360] [AsyncOrchestrator] All stages initialized successfully
WARNING 02-24 09:59:08 [async_omni.py:274] [AsyncOrchestrator] No LLM stage found, processors will not be available. This may cause issues with OpenAIServingModels.
INFO 02-24 09:59:08 [omni_stage.py:566] [Stage-0] Status transitioned to: TRANSITIONING
INFO 02-24 09:59:08 [omni_stage.py:570] [Stage-0] Submitting SLEEP task (Level: 2)
INFO 02-24 09:59:08 [async_omni.py:894] [AsyncOrchestrator] Sleep initiated. Awaiting confirmation from 1 workers...
[Stage-0] INFO 02-24 09:59:08 [async_omni_diffusion.py:309] [Entrypoint] Relaying Sleep Task: 05eda0a7-27f5-4843-ad1a-24291b0dc7d4 (Level: 2)
[Stage-0] INFO 02-24 09:59:08 [diffusion_engine.py:401] [Diffusion Engine Relay] Dispatching Sleep Task 05eda0a7-27f5-4843-ad1a-24291b0dc7d4 (Level: 2)
[Stage-0] INFO 02-24 09:59:08 [diffusion_worker.py:221] [Diffusion Worker 0] Handshake Received: Task 05eda0a7-27f5-4843-ad1a-24291b0dc7d4, Level 2
[Stage-0] INFO 02-24 09:59:08 [cumem.py:213] CuMemAllocator: sleep freed 14.89 GiB memory in total, of which 0.00 GiB is backed up in CPU and the rest 14.89 GiB is discarded directly.
[Stage-0] INFO 02-24 09:59:09 [diffusion_worker.py:198] [Diffusion Worker 0] Level 2 Sleep: Freed 18.40 GiB. 0.62GiB memory is still in use.
[Stage-0] INFO 02-24 09:59:09 [diffusion_worker.py:239] [Diffusion Worker 0]: ACK emitted. Freed 18.40 GiB.
[Stage-0] INFO 02-24 09:59:09 [omni_stage.py:1323] [Stage-0] Sleep ACKs forwarded to Orchestrator
INFO 02-24 09:59:09 [async_omni.py:641] [AsyncOrchestrator] Intercepted wrapped ACK for task 05eda0a7-27f5-4843-ad1a-24291b0dc7d4 from stage-0
INFO 02-24 09:59:09 [omni_stage.py:566] [Stage-0] Status transitioned to: SLEEPING
INFO 02-24 09:59:10 [omni_stage.py:579] [Stage-0] Submitting WAKE_UP task
INFO 02-24 09:59:10 [async_omni.py:921] [AsyncOrchestrator] Wake-up initiated. Awaiting confirmation from 1 workers...
[Stage-0] INFO 02-24 09:59:10 [async_omni_diffusion.py:318] [Entrypoint] Relaying WakeUp Task: c9a8e0a2-699a-4f3c-ac62-6b7bb08b3b43
[Stage-0] INFO 02-24 09:59:10 [diffusion_engine.py:424] [Diffusion Engine Relay] Dispatching Wake-up Task c9a8e0a2-699a-4f3c-ac62-6b7bb08b3b43 to workers...
[Stage-0] INFO 02-24 09:59:10 [diffusion_worker.py:250] [Diffusion Worker 0] Responding to Wake-up Task: c9a8e0a2-699a-4f3c-ac62-6b7bb08b3b43
[Stage-0] INFO 02-24 09:59:10 [diffusion_worker.py:210] [Diffusion Worker 0] Wake-up complete.
[Stage-0] INFO 02-24 09:59:10 [diffusion_worker.py:259] [Diffusion Worker 0] Wake-up confirmed.
INFO 02-24 09:59:10 [async_omni.py:641] [AsyncOrchestrator] Intercepted wrapped ACK for task c9a8e0a2-699a-4f3c-ac62-6b7bb08b3b43 from stage-0
INFO 02-24 09:59:10 [omni_stage.py:566] [Stage-0] Status transitioned to: RUNNING
INFO 02-24 09:59:11 [async_omni.py:378] [AsyncOrchestrator] Entering scheduling loop: stages=1, final_stage=0
[Stage-0] INFO 02-24 09:59:11 [manager.py:538] Deactivating all adapters: 0 layers
[Stage-0] WARNING 02-24 09:59:11 [kv_transfer_manager.py:356] No connector available for receiving KV cache
[Stage-0] INFO 02-24 09:59:11 [diffusion_engine.py:92] Generation completed successfully.
[Stage-0] INFO 02-24 09:59:11 [diffusion_engine.py:110] Post-processing completed in 0.0583 seconds
PASSED[Stage-0] INFO 02-24 09:59:26 [diffusion_worker.py:400] Worker 0: Received shutdown message
[Stage-0] INFO 02-24 09:59:26 [diffusion_worker.py:421] event loop terminated.
[Stage-0] INFO 02-24 09:59:26 [diffusion_worker.py:452] Worker 0: Shutdown complete.
[Stage-0] INFO 02-24 09:59:29 [async_omni_diffusion.py:213] AsyncOmniDiffusion closed
[Stage-0] INFO 02-24 09:59:29 [omni_stage.py:1428] Stage worker exiting
GPU cleanup disabled


=============================== warnings summary ===============================
<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute

<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyObject has no __module__ attribute

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================== 6 passed, 2 warnings in 309.81s (0:05:09) ===================

```

AMD MI300X TP = 2 （Core pytest test output）

llm and generation worker Skipping on AMD. To be placed in the next plan

```
tests/entrypoints/test_omni_sleep_mode.py::TestOmniSleepMode::test_diffusion_vram_lifecycle_audit 
=== PRE-TEST GPU CLEANUP ===
GPU cleanup disabled
INFO 02-24 10:18:53 [scheduler.py:224] Chunked prefill is enabled with max_num_batched_tokens=2048.
--- Running test: test_diffusion_vram_lifecycle_audit
INFO 02-24 10:18:53 [omni.py:117] Initializing stages for model: black-forest-labs/FLUX.2-klein-4B
.................................................................................
[Stage-0] INFO 02-24 10:19:48 [async_omni_diffusion.py:113] AsyncOmniDiffusion initialized with model: black-forest-labs/FLUX.2-klein-4B
INFO 02-24 10:19:48 [omni.py:334] [AsyncOrchestrator] Stage-0 reported ready
INFO 02-24 10:19:48 [omni.py:360] [AsyncOrchestrator] All stages initialized successfully
WARNING 02-24 10:19:48 [async_omni.py:274] [AsyncOrchestrator] No LLM stage found, processors will not be available. This may cause issues with OpenAIServingModels.
INFO 02-24 10:19:49 [omni_stage.py:566] [Stage-0] Status transitioned to: TRANSITIONING
INFO 02-24 10:19:49 [omni_stage.py:570] [Stage-0] Submitting SLEEP task (Level: 2)
INFO 02-24 10:19:49 [async_omni.py:894] [AsyncOrchestrator] Sleep initiated. Awaiting confirmation from 1 workers...
[Stage-0] INFO 02-24 10:19:49 [async_omni_diffusion.py:309] [Entrypoint] Relaying Sleep Task: 7e306879-32c1-4780-8f29-d5dd8339b15a (Level: 2)
[Stage-0] INFO 02-24 10:19:49 [diffusion_engine.py:401] [Diffusion Engine Relay] Dispatching Sleep Task 7e306879-32c1-4780-8f29-d5dd8339b15a (Level: 2)
[Stage-0] INFO 02-24 10:19:49 [diffusion_worker.py:221] [Diffusion Worker 0] Handshake Received: Task 7e306879-32c1-4780-8f29-d5dd8339b15a, Level 2
[Stage-0] INFO 02-24 10:19:49 [cumem.py:213] CuMemAllocator: sleep freed 14.88 GiB memory in total, of which 0.00 GiB is backed up in CPU and the rest 14.88 GiB is discarded directly.
[Stage-0] INFO 02-24 10:19:49 [diffusion_worker.py:198] [Diffusion Worker 0] Level 2 Sleep: Freed 15.90 GiB. 2.06GiB memory is still in use.
[Stage-0] INFO 02-24 10:19:49 [diffusion_worker.py:239] [Diffusion Worker 0]: ACK emitted. Freed 15.90 GiB.
[Stage-0] INFO 02-24 10:19:49 [omni_stage.py:1323] [Stage-0] Sleep ACKs forwarded to Orchestrator
INFO 02-24 10:19:49 [async_omni.py:641] [AsyncOrchestrator] Intercepted wrapped ACK for task 7e306879-32c1-4780-8f29-d5dd8339b15a from stage-0
INFO 02-24 10:19:49 [omni_stage.py:566] [Stage-0] Status transitioned to: SLEEPING
INFO 02-24 10:19:50 [omni_stage.py:579] [Stage-0] Submitting WAKE_UP task
INFO 02-24 10:19:50 [async_omni.py:921] [AsyncOrchestrator] Wake-up initiated. Awaiting confirmation from 1 workers...
[Stage-0] INFO 02-24 10:19:50 [async_omni_diffusion.py:318] [Entrypoint] Relaying WakeUp Task: 7b92d7e7-ea5d-400b-91d3-fc5008f92ea5
[Stage-0] INFO 02-24 10:19:50 [diffusion_engine.py:424] [Diffusion Engine Relay] Dispatching Wake-up Task 7b92d7e7-ea5d-400b-91d3-fc5008f92ea5 to workers...
[Stage-0] INFO 02-24 10:19:50 [diffusion_worker.py:250] [Diffusion Worker 0] Responding to Wake-up Task: 7b92d7e7-ea5d-400b-91d3-fc5008f92ea5
[Stage-0] INFO 02-24 10:19:50 [diffusion_worker.py:210] [Diffusion Worker 0] Wake-up complete.
[Stage-0] INFO 02-24 10:19:50 [diffusion_worker.py:259] [Diffusion Worker 0] Wake-up confirmed.
INFO 02-24 10:19:50 [async_omni.py:641] [AsyncOrchestrator] Intercepted wrapped ACK for task 7b92d7e7-ea5d-400b-91d3-fc5008f92ea5 from stage-0
INFO 02-24 10:19:50 [omni_stage.py:566] [Stage-0] Status transitioned to: RUNNING
INFO 02-24 10:19:51 [async_omni.py:378] [AsyncOrchestrator] Entering scheduling loop: stages=1, final_stage=0
[Stage-0] INFO 02-24 10:19:51 [manager.py:538] Deactivating all adapters: 0 layers
[Stage-0] WARNING 02-24 10:19:51 [kv_transfer_manager.py:356] No connector available for receiving KV cache
[Stage-0] INFO 02-24 10:19:51 [diffusion_engine.py:92] Generation completed successfully.
[Stage-0] INFO 02-24 10:19:52 [diffusion_engine.py:110] Post-processing completed in 0.1198 seconds
PASSED[Stage-0] INFO 02-24 10:20:07 [diffusion_worker.py:400] Worker 0: Received shutdown message
[Stage-0] INFO 02-24 10:20:07 [diffusion_worker.py:421] event loop terminated.
[Stage-0] INFO 02-24 10:20:07 [diffusion_worker.py:452] Worker 0: Shutdown complete.
[Stage-0] INFO 02-24 10:20:09 [async_omni_diffusion.py:213] AsyncOmniDiffusion closed
[Stage-0] INFO 02-24 10:20:09 [omni_stage.py:1428] Stage worker exiting
GPU cleanup disabled


=============================== warnings summary ===============================
<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute

<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyObject has no __module__ attribute

../../usr/local/lib/python3.12/dist-packages/antlr4/Lexer.py:12
  /usr/local/lib/python3.12/dist-packages/antlr4/Lexer.py:12: DeprecationWarning: typing.io is deprecated, import directly from typing instead. typing.io will be removed in Python 3.13.
    from typing.io import TextIO

../../usr/local/lib/python3.12/dist-packages/antlr4/Parser.py:6
  /usr/local/lib/python3.12/dist-packages/antlr4/Parser.py:6: DeprecationWarning: typing.io is deprecated, import directly from typing instead. typing.io will be removed in Python 3.13.
    from typing.io import TextIO

tests/entrypoints/test_omni_sleep_mode.py::TestOmniSleepMode::test_diffusion_integrity_bit_level
tests/entrypoints/test_omni_sleep_mode.py::TestOmniSleepMode::test_diffusion_integrity_bit_level
tests/entrypoints/test_omni_sleep_mode.py::TestOmniSleepMode::test_diffusion_vram_lifecycle_audit
  /app/vllm-omni/vllm_omni/distributed/omni_connectors/utils/serialization.py:290: DeprecationWarning: 'mode' parameter is deprecated and will be removed in Pillow 13 (2026-10-15)
    return Image.fromarray(arr, mode=mode)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
============= 4 passed, 2 skipped, 7 warnings in 267.62s (0:04:27) =============
sys:1: DeprecationWarning: builtin type swigvarlink has no __module__ attribute

```

2. 
[A6000] Diffusion VRAM Lifecycle

vram_unitest_6

Observation: Successful reclamation of ~16.8GB. Wakeup restores the engine to a cleaner steady state.

[MI300X] Diffusion VRAM Lifecycle

vram_unitest_6

Observation: Despite ROCm's aggressive pre-allocation (28.6GB peak), Level 2 Sleep purges all compile-time workspaces, returning to a 17.8GB compact footprint after wakeup.

### CODE REVIEW

File: vllm_omni/entrypoints/omni_stage.py [hunk 1]
```diff
@@ -295,6 +295,7 @@ def __init__(self, stage_config: Any, stage_init_timeout: int = 300):
         self._proc: mp.Process | None = None
         self._shm_threshold_bytes: int = 65536
         self._stage_init_timeout: int = stage_init_timeout
+        self.status = " RUNNING"
```
Comment (lishunyang12) [inline vllm_omni/entrypoints/omni_stage.py:299]: `self.status = " RUNNING"` — leading space will break status checks like `stage.status == "SLEEPING"` later. Should be `"RUNNING"`.

File: vllm_omni/worker/base.py [hunk 2]
```diff
@@ -106,3 +117,100 @@ def determine_available_memory(self) -> int:
...
+        """
+        Put the worker to sleep.
+        Args:
+            level: 1 (Offload weights to CPU), level: 2 (Total Discard).
+        """
+        from vllm.device_allocator.cumem import CuMemAllocator
+        mem_before = current_omni_platform.get_current_memory_usage(self.device)
+        offload_tags = ("weights") if level == 1 else tuple()
+        allocator = CuMemAllocator.get_instance()
+        allocator.sleep(offload_tags=offload_tags)
```
Comment (lishunyang12) [inline vllm_omni/worker/base.py:148]:
```suggestion
        offload_tags = ("weights",) if level == 1 else tuple()
```
Missing trailing comma — `("weights")` is just `"weights"`, not a tuple.

File: vllm_omni/entrypoints/async_omni.py [hunk 3]
```diff
@@ -65,6 +66,54 @@ def _weak_close_cleanup_async(stage_list, stage_in_queues, stage_out_queues, ray
...
+        fut = loop.create_future()
+        self._pending_tasks[task_id] = {
+            "future": fut,
+            "expected_count": expected_count,
+            "current_count": 0
+        }
+        return fut
+    async def resolve(self, ack: OmniACK):
+        # ACK for dictionary type
+        tid = getattr(ack, "task_id", ack.get("task_id") if isinstance(ack, dict) else None)
```
Comment (lishunyang12) [inline vllm_omni/entrypoints/async_omni.py:92]: Method name mismatch: `record_vram_reclaimed` is called here but the actual method in `OmniStatsCollector` is named `record_vram_reclamation`. This will silently fail (caught by the bare `except`).

File: vllm_omni/config/model.py [hunk 4]
```diff
@@ -96,22 +100,93 @@ def draw_hf_text_config(self):
...
+        enable_mm_embeds      = extract_var(1, "enable_mm_embeds")
+        media_io_kwargs       = extract_var(2, "media_io_kwargs")
+        mm_processor_kwargs   = extract_var(3, "mm_processor_kwargs")
+        mm_processor_cache_gb = extract_var(4, "mm_processor_cache_gb")
+        mm_processor_cache_type = extract_var(5, "mm_processor_cache_type")
+        mm_shm_cache_max_object_size_mb = extract_var(6, "mm_shm_cache_max_object_size_mb")
+        mm_encoder_only       = extract_var(7, "mm_encoder_only", False)
+        mm_encoder_tp_mode    = extract_var(8, "mm_encoder_tp_mode", "weights")
+        if mm_encoder_tp_mode not in ["weights", "data"]:
+            mm_encoder_tp_mode = "weights"
```
Comment (lishunyang12) [inline vllm_omni/config/model.py:125]: The bulk of this `__post_init__` (lines ~100-220 added here) duplicates logic from `ModelConfig.__post_init__` verbatim — model redirect, hf_config loading, dtype verification, max_model_len, pooler config, multimodal config init, etc. This will silently drift out of sync with upstream vllm whenever `ModelConfig` changes.

Can you call `super().__post_init__(...)` for the shared logic and only keep the omni-specific overrides (stage_connector_config, draw_hf_text_config) here?

File: tests/entrypoints/test_omni_sleep_mode.py [hunk 5]
```diff
@@ -0,0 +1,240 @@
+import asyncio
+import torch
+is_rocm = torch.version.hip is not None
+import pytest
+import torch
```
Comment (lishunyang12) [inline tests/entrypoints/test_omni_sleep_mode.py:5]: Duplicate `import torch` (lines 2 and 4). Also `uuid` is imported but only used in the test fixtures — minor, but pre-commit should catch the duplicate.

**Review by lishunyang12 — COMMENTED** [review]
Left a few comments. The main concern is the \ copying ~120 lines from upstream \ — that's going to be a maintenance headache on every vllm rebase. The status string typo and tuple bug in \ are also real issues.

**Review by lishunyang12 — COMMENTED** [review]
Left a few comments. The main concern is the `OmniModelConfig.__post_init__` copying ~120 lines from upstream `ModelConfig` — that's going to be a maintenance headache on every vllm rebase. The status string typo and tuple bug in `base.py` are also real issues.

File: vllm_omni/entrypoints/omni_stage.py (hunk 1 above)
Comment (Flink-ddd) [reply vllm_omni/entrypoints/omni_stage.py:299]: yes, Thanks for your quickly reply, I will update these code logic in new open PR soon.

File: vllm_omni/entrypoints/async_omni.py
Comment (lrq619) [inline vllm_omni/entrypoints/async_omni.py:233]: Need to add `enable_sleep_mode": kwargs.get("enable_sleep_mode", False),`

**Review by lrq619 — COMMENTED** [review]
Hi thanks for the great PR, I have tested it and it worked after some minor modifications, I will put my modifications here for your reference.

1. In `entrypoints/async_omni.py` need to add a line to pass "--enable_sleep_mode"
2. In `entrypoints/utils.py`, I append the following codes at the end of `build_base_engine_args`
```
    # Extract shared overrides that should apply to all stages when provided.
    shared_keys = [
        "enable_sleep_mode",
    ]
    if is_dict_like:
        shared_overrides = {k: source[k] for k in shared_keys if k in source and source[k] is not None}
    else:
        shared_overrides = {
            k: getattr(source, k) for k in shared_keys if hasattr(source, k) and getattr(source, k) is not None
        }

    if shared_overrides:
        base_engine_args = base_engine_args or {}
        base_engine_args.update(shared_overrides)
```

File: vllm_omni/entrypoints/async_omni.py
Comment (Flink-ddd) [reply vllm_omni/entrypoints/async_omni.py:233]: hi @lrq619 , Thanks for your test work and suggestion,  I will update these code logic in new open PR soon.  the maintainer suggested merging this feature in version 0.18, so I will open new PR base on latest main code.

**Comment by Gaohan123** [comment]
Hello, any updates? Is it still in draft status?

**Comment by Flink-ddd** [comment]
Hi @Gaohan123 , I need according to latest code to update draft PR code and re-run all pytest test, next week I will submit a new PR. Thanks

**Comment by lishunyang12** [comment]
Closed as the PR has been superseded.
